code_generator/GdkCodeGenerator/Templates/UnityGameObjectComponentDispatcherGenerator.tt

```diff
@@ -20,58 +20,106 @@ namespace <#= qualifiedNamespace #>
             {
 <# if (fieldDetailsList.Count > 0) { #>
-                typeof(ComponentsUpdated<<#= componentDetails.TypeName #>.Update>), typeof(GameObjectReference)
+                ComponentType.ReadOnly<ComponentsUpdated<<#= componentDetails.TypeName #>.Update>>(), ComponentType.ReadOnly<GameObjectReference>()
 <# } #>
             };
 
             public override ComponentType[][] EventsReceivedComponentTypeArrays => new ComponentType[][]
             {
 <# foreach(var eventDetails in eventDetailsList) { #>
-                new ComponentType[] { typeof(EventsReceived<<#= eventDetails.EventName #>Event>), typeof(GameObjectReference) },
+                new ComponentType[] { ComponentType.ReadOnly<EventsReceived<<#= eventDetails.EventName #>Event>>(), ComponentType.ReadOnly<GameObjectReference>() },
 <# } #>
             };
 
             public override ComponentType[][] CommandRequestsComponentTypeArrays => new ComponentType[][]
             {
 <# foreach(var commandDetails in commandDetailsList) { #>
-                new ComponentType[] { typeof(CommandRequests<<#= commandDetails.CommandName #>.Request>), typeof(GameObjectReference) },
+                new ComponentType[] { ComponentType.ReadOnly<CommandRequests<<#= commandDetails.CommandName #>.Request>>(), ComponentType.ReadOnly<GameObjectReference>() },
 <# } #>
             };
 
-            public override void InvokeOnAddComponentLifecycleCallbacks(GameObjectDispatcherSystem gameObjectDispatcherSystem)
+            protected override uint GetComponentId()
             {
-                var entities = ComponentAddedComponentGroup.GetEntityArray();
+                return <#= unityComponentDefinition.Id #>;
+            }
+
+            public override void InvokeOnComponentUpdateUserCallbacks(GameObjectDispatcherSystem gameObjectDispatcherSystem)
+            {
+<# if (fieldDetailsList.Count > 0) { #>
+                var entities = ComponentsUpdatedComponentGroup.GetEntityArray();
+                var updateLists = ComponentsUpdatedComponentGroup.GetComponentArray<ComponentsUpdated<<#= componentDetails.TypeName #>.Update>>();
                 for (var i = 0; i < entities.Length; i++)
                 {
-                    var spatialOSBehaviourManager = gameObjectDispatcherSystem.GetSpatialOSBehaviourManager(entities[i].Index);
-                    spatialOSBehaviourManager.AddComponent(<#= unityComponentDefinition.Id #>);
+                    var readers = gameObjectDispatcherSystem.GetSpatialOSBehaviourManager(entities[i].Index)
+                        .GetReadersWriters(<#= unityComponentDefinition.Id #>);
+                    var updateList = updateLists[i];
+                    foreach (var reader in readers)
+                    {
+                        foreach (var update in updateList.Buffer)
+                        {
+                            ((<#= componentDetails.TypeName #>.ReaderWriterImpl) reader).OnComponentUpdate(update);
```

Comment:
Prefer `foreach(var reader in readers.OfType<<#= componentDetails.TypeName #>.ReaderWriterImpl>` to shift the cast out (here and in the other cases.)

Reply:
Good idea, is it OK if I just keep it in mind and driveby change whenever I next touch this file? Probably will happen soon because there's things to do regarding order of operations.

Reply:
Yep, that's fine